### DISCUSSION TITLE
Can I use `TypeVar` and `TypeAdapter` together?

### DISCUSSION THREAD

**Discussion by vadim-su** [pydantic/pydantic#5766] [Pydantic V2]
I often have to convert a list of different models and I would like to do it somehow through `TypeAdaper` and `TypeVar`.

```py
TypeToListVar = TypeVar('TypeToListVar')
ModelListType = TypeAdapter(list[TypeToListVar])
```

Can I do something like that?
`ModelsListType[VoiceRegion].validate_python(resp.body)`

**Comment by dmontagu** [comment]
Currently, no. I think the best and most performant way to achieve something like this would be to do something like:

```python
from functools import lru_cache
from typing import TypeVar

from pydantic import TypeAdapter

T = TypeVar("T")

@lru_cache
def models_list_type(type_: type[T]) -> TypeAdapter[list[T]]:
    return TypeAdapter(list[type_])  # type: ignore
```

The `lru_cache` means you'll only pay the price of creating the type adapter once per model, and
```python
ModelsListType[VoiceRegion].validate_python(resp.body)
```
just becomes
```python
models_list_type(VoiceRegion).validate_python(resp.body)
```
(Note I haven't tested the code above, but I think it should work.)

**Reply by vadim-su** [reply]
Thanks, It works. But a type checker reports that the output types do not match.
```py

@lru_cache
def models_list_type(type_: type[T]) -> TypeAdapter[list[T]]:
    return TypeAdapter(list[type_])  # type: ignore

var: list[VoiceRegion] =  models_list_type(VoiceRegion).validate_python(resp.body) # <<
```

```
Expression of type "list[T@models_list_type]" cannot be assigned to declared type "list[VoiceRegion]"
  "list[T@models_list_type]" is incompatible with "list[VoiceRegion]"
    TypeVar "_T@list" is invariant
      Type "T@models_list_type" cannot be assigned to type "VoiceRegion"
```

**Reply by dmontagu** [reply]
This looks to me like a bug in pyright, which I intend to report shortly; if you drop the `@lru_cache` that issue seems to go away.

You can work around this via:
```python
def models_list_type(type_: type[T]) -> TypeAdapter[list[T]]:
    return TypeAdapter(list[type_])  # type: ignore

if not typing.TYPE_CHECKING:
    models_list_type = lru_cache(models_list_type)
```
that should pass type checking and have the same logic. (And be correct.)

**Reply by dmontagu** [reply]
https://github.com/microsoft/pyright/discussions/5123 for reference
